File: players/dancer/planner.py
"""L2 DuelPlanner: exact-model search for the 1v1 melee game.

The melee response is deterministic (verified 33/33): attack at
Manhattan-1 on its own turn, else one step reducing the strictly
largest gap axis (tie -> column), stalling when blocked. Players act
first within a tick. So the local duel is perfectly simulable: search
our action tree DEPTH plies deep with the exact enemy response and play
the best line. The search rediscovers the diagonal sword dance, the
safe-stall escape, first-strike holds, corner play and trade decisions
from the rules alone.

Intent comes from the executive:
  kill  - low damage aversion, kill bonus dominates
  avoid - high damage aversion, distance shaping dominates

Occupancy cells (other tracks) are soft penalties, never hard walls:
phantom occupancy must not cage the search (measured v3 failure).
"""

from __future__ import annotations

from ..scripted_player import (ATN_ATTACK, ATN_NOOP, MOVE_DELTAS,
                               RUN_OFFSET)

DEPTH = 5
KILL_BONUS = 10.0     # small closure bonus; the real reward is per-hit
HIT_CREDIT = 0.6      # value per point of damage dealt - proportional
                      # credit keeps partial progress visible inside the
                      # horizon (a flat kill bonus made tanking beat the
                      # dance: only the tank-kill fits in 5 plies)
OCC_PENALTY = 2.5
W_KILL = 5.0          # per-hit-taken cost > per-hit credit: prefer
                      # clean lines; free-hit dances still land kills
W_AVOID = 9.0


def plan_duel(our_pos: tuple[int, int],
              enemy_pos: tuple[int, int],
              enemy_hp: int,
              our_dmg: int,
              their_dmg: int,
              sword: bool,
              intent_kill: bool,
              passable,          # (r, c) -> bool  (terrain, window-clamped)
              occupied: set[tuple[int, int]] = frozenset(),
              depth: int = DEPTH) -> int:
    """Best action (int) for the duel state. Coordinates are any
    consistent integer frame (window cells work)."""
    W = W_KILL if intent_kill and our_dmg > 0 else W_AVOID
    dmg = max(our_dmg, 0)

    def in_arc(pr, pc, r, c):
        ar, ac = abs(r - pr), abs(c - pc)
        if ar + ac == 1:
            return True
        return sword and ar == 1 and ac == 1

    def enemy_step(pr, pc, r, c):
        """(nr, nc, attacked) exact chase response."""
        dr, dc = pr - r, pc - c
        if abs(dr) + abs(dc) == 1:
            return r, c, True
        if abs(dr) > abs(dc):
            nr, nc = r + (1 if dr > 0 else -1), c
        else:
            nr, nc = r, c + (1 if dc > 0 else -1)
        if passable(nr, nc) and (nr, nc) != (pr, pc) \
                and (nr, nc) not in occupied:
            return nr, nc, False
        return r, c, False

    def leaf_value(pr, pc, r, c):
        d = max(abs(r - pr), abs(c - pc))
        return d * (0.8 if not intent_kill else 0.05)

    def search(pr, pc, r, c, ehp, d):
        if ehp <= 0:
            return KILL_BONUS + d * 2.0, ATN_NOOP  # already credited
        if d == 0:
            return leaf_value(pr, pc, r, c), ATN_NOOP
        best_v, best_a = -1e18, ATN_NOOP
        cands = [(ATN_NOOP, pr, pc, False)]
        if dmg > 0 and in_arc(pr, pc, r, c):
            cands.append((ATN_ATTACK, pr, pc, True))
        for atn, (mr, mc) in MOVE_DELTAS.items():
            nr, nc = pr + mr, pc + mc
            if passable(nr, nc) and (nr, nc) != (r, c):
                cands.append((atn, nr, nc, False))
                n2r, n2c = pr + 2 * mr, pc + 2 * mc
                if passable(n2r, n2c) and (n2r, n2c) != (r, c) \
                        and (nr, nc) not in occupied:
                    cands.append((atn + RUN_OFFSET, n2r, n2c, False))
        for atn, npr, npc, attacked in cands:
            nehp = ehp - dmg if attacked else ehp
            if nehp <= 0:
                val = KILL_BONUS + dmg * HIT_CREDIT + d * 2.0
            else:
                ner, nec, ehit = enemy_step(npr, npc, r, c)
                sub, _ = search(npr, npc, ner, nec, nehp, d - 1)
                val = sub + (dmg * HIT_CREDIT if attacked else 0.0) \
                    - (W * their_dmg / 10.0 if ehit else 0.0)
            if (npr, npc) != (pr, pc) and (npr, npc) in occupied:
                val -= OCC_PENALTY
            if val > best_v:
                best_v, best_a = val, atn
        return best_v, best_a

    _v, act = search(our_pos[0], our_pos[1],
                     enemy_pos[0], enemy_pos[1], enemy_hp, depth)
    return act

File: players/nimagent/planner.nim
# L2 planner: exact-model search for the local melee game.
# Port of players/dancer/planner.py plan_melee (the N-enemy search that
# passed every MiniMelee gate and stress variant on 2026-08-03).
#
# Verified rules: players act first within a tick; each melee attacks
# at Manhattan-1 on its turn, else steps 1 along the strictly-largest
# gap axis (tie -> column), stalling when blocked by terrain, the
# player, or another enemy; enemy-enemy collisions block (a cell
# vacated earlier the same tick is free).

import percept

const
  Depth* = 5
  KillBonus = 10.0
  HitCredit = 0.6
  OccPenalty = 2.5
  HazardPenalty = 14.0
  WKill = 5.0
  WAvoid = 9.0
  MaxEnemies* = 8

type
  PlanEnemy* = object
    r*, c*: int
    dmg*: float          # their damage to us per hit
    hp*: int
    ranged*: bool        # bow enemy: shoots when aligned <=4, else
                         # steps along the SHORTEST axis toward
                         # alignment; NO melee attack (enemy_ai)

  Cell = tuple[r, c: int]

  PlanCtx = object
    enemies: seq[PlanEnemy]
    targetIdx: int       # -1 = pure escape
    dmg: float           # our damage per hit (0 = cannot hurt)
    sword: bool
    bow: bool            # held bow: attack arc is ON-AXIS <= 4
                         # (ATTACK_BOW, nmmo3.h:891) - the kite weapon
    intentKill: bool
    w: float
    hp: float            # our hp: cumulative damage >= hp is DEATH,
                         # not a recoverable cost
    passable: proc (r, c: int): bool {.closure.}
    occupied: seq[Cell]
    hazards: seq[Cell]

proc inArc(ctx: PlanCtx, pr, pc, r, c: int): bool =
  let ar = abs(r - pr)
  let ac = abs(c - pc)
  if ctx.bow:
    return (ar == 0 or ac == 0) and max(ar, ac) <= 4 and ar + ac > 0
  if ar + ac == 1: return true
  ctx.sword and ar == 1 and ac == 1

proc stepEnemies(ctx: PlanCtx, pr, pc: int, epos: var seq[Cell],
                 alive: seq[bool]): float =
  ## All live enemies act vs player at (pr,pc): returns damage taken;
  ## mutates epos in place. List order; vacated cells free up.
  ## AGGRO CUTOFF (enemy_ai, nmmo3.h:1592): an enemy only acts when a
  ## player is inside its +-4 box - at Chebyshev >= 5 it reverts to
  ## leashed wandering (modeled as a stall). Escape is absolute.
  var occ: seq[Cell]
  for i in 0 ..< epos.len:
    if alive[i]: occ.add epos[i]
  for i in 0 ..< epos.len:
    if not alive[i]: continue
    let (r, c) = epos[i]
    let dr = pr - r
    let dc = pc - c
    if max(abs(dr), abs(dc)) > 4:
      continue                       # outside its aggro box: idle
    var nr = r
    var nc = c
    if ctx.enemies[i].ranged:
      # bow: shoot when aligned (incl. adjacent); else step along the
      # SHORTEST axis toward alignment. No melee attack.
      if dr == 0 or dc == 0:
        result += ctx.enemies[i].dmg
        continue
      if abs(dr) > abs(dc):
        nc = c + (if dc > 0: 1 else: -1)
      else:
        nr = r + (if dr > 0: 1 else: -1)
    else:
      if abs(dr) + abs(dc) == 1:
        result += ctx.enemies[i].dmg
        continue
      if abs(dr) > abs(dc):
        nr = r + (if dr > 0: 1 else: -1)
      else:
        nc = c + (if dc > 0: 1 else: -1)
    if ctx.passable(nr, nc) and (nr, nc) != (pr, pc) and
        (nr, nc) notin occ:
      let idx = occ.find((r, c))
      if idx >= 0: occ.del(idx)
      occ.add (nr, nc)
      epos[i] = (nr, nc)

proc leafValue(ctx: PlanCtx, pr, pc: int, epos: seq[Cell],
               alive: seq[bool]): float =
  var d = 9
  for i in 0 ..< epos.len:
    if alive[i]:
      d = min(d, max(abs(epos[i].r - pr), abs(epos[i].c - pc)))
  result = float(d) * (if ctx.intentKill: 0.05 else: 0.8)
  if d >= 5 and not ctx.intentKill:
    result += 2.0       # outside every aggro box: fully disengaged

proc search(ctx: PlanCtx, pr, pc: int, epos: seq[Cell],
            alive: seq[bool], thp: float, d: int,
            hpLeft: float): tuple[v: float, a: int] =
  if ctx.targetIdx >= 0 and thp <= 0:
    return (KillBonus + float(d) * 2.0, AtnNoop)
  if d == 0:
    return (ctx.leafValue(pr, pc, epos, alive), AtnNoop)
  var bestV = -1e18
  var bestA = AtnNoop
  # candidates: (action, npr, npc, attacked)
  var cands: seq[tuple[a, npr, npc: int, attacked: bool]]
  cands.add (AtnNoop, pr, pc, false)
  if ctx.dmg > 0 and ctx.targetIdx >= 0 and alive[ctx.targetIdx] and
      ctx.inArc(pr, pc, epos[ctx.targetIdx].r, epos[ctx.targetIdx].c):
    cands.add (AtnAttack, pr, pc, true)
  for ai in 0 .. 3:
    let (mr, mc) = MoveDeltas[ai]
    let nr = pr + mr
    let nc = pc + mc
    var blockedByEnemy = false
    for i in 0 ..< epos.len:
      if alive[i] and epos[i] == (nr, nc): blockedByEnemy = true; break
    if ctx.passable(nr, nc) and not blockedByEnemy:
      cands.add (ai, nr, nc, false)
      if mr == 0:
        # horizontal dance runs only (vertical runs can outrun the
        # shallow +-5-row view)
        let n2r = pr + 2 * mr
        let n2c = pc + 2 * mc
        var blocked2 = false
        for i in 0 ..< epos.len:
          if alive[i] and epos[i] == (n2r, n2c): blocked2 = true; break
        if ctx.passable(n2r, n2c) and not blocked2 and
            (nr, nc) notin ctx.occupied:
          cands.add (ai + RunOffset, n2r, n2c, false)
  for cand in cands:
    var val: float
    let nthp = if cand.attacked: thp - ctx.dmg else: thp
    if ctx.targetIdx >= 0 and nthp <= 0:
      val = KillBonus + ctx.dmg * HitCredit + float(d) * 2.0
    else:
      var nepos = epos
      let taken = ctx.stepEnemies(cand.npr, cand.npc, nepos, alive)
      let nhpLeft = hpLeft - taken
      if nhpLeft <= 0:
        # this line kills us: terminal, catastrophic, sooner is worse
        val = -1000.0 - float(d)
      else:
        let sub = ctx.search(cand.npr, cand.npc, nepos, alive, nthp,
                             d - 1, nhpLeft)
        val = sub.v +
          (if cand.attacked: ctx.dmg * HitCredit else: 0.0) -
          ctx.w * taken / 10.0
    if (cand.npr, cand.npc) != (pr, pc):
      if (cand.npr, cand.npc) in ctx.occupied: val -= OccPenalty
      if (cand.npr, cand.npc) in ctx.hazards: val -= HazardPenalty
    if val > bestV:
      bestV = val
      bestA = cand.a
  (bestV, bestA)

proc planMelee*(ourR, ourC: int,
                enemies: seq[PlanEnemy],
                targetIdx: int,          # -1 = pure escape
                ourDmg: float,
                sword: bool,
                intentKill: bool,
                passable: proc (r, c: int): bool {.closure.},
                occupied: seq[Cell] = @[],
                hazards: seq[Cell] = @[],
                depth = Depth,
                ourHp = 99.0,
                bowHeld = false): int =
  ## Best action for the local melee state. Any consistent integer
  ## frame (window cells work).
  var ctx = PlanCtx(
    enemies: enemies, targetIdx: targetIdx,
    dmg: max(ourDmg, 0.0), sword: sword, intentKill: intentKill,
    w: (if intentKill and ourDmg > 0: WKill else: WAvoid),
    bow: bowHeld,
    hp: ourHp,
    passable: passable, occupied: occupied, hazards: hazards)
  var epos: seq[Cell]
  var alive: seq[bool]
  for e in enemies:
    epos.add (e.r, e.c)
    alive.add true
  let thp = if targetIdx >= 0: float(enemies[targetIdx].hp) else: 1.0
  ctx.search(ourR, ourC, epos, alive, thp, depth, ourHp).a
